React/Views/RCTNavItem.m

```diff
@@ -11,75 +11,139 @@
 
 @implementation RCTNavItem
 
-@synthesize backButtonItem = _backButtonItem;
-@synthesize leftButtonItem = _leftButtonItem;
-@synthesize rightButtonItem = _rightButtonItem;
+- (void)setOnNavLeftButtonTap:(RCTBubblingEventBlock)onNavLeftButtonTap
+{
+  _onNavLeftButtonTap = onNavLeftButtonTap;
+}
+
+- (void)setNavigationBar:(UINavigationBar *)navigationBar
+{
+  _navigationBar = navigationBar;
+  _navigationBar.barTintColor = self.barTintColor;
+  _navigationBar.tintColor = self.tintColor;
+  _navigationBar.translucent = self.translucent;
+  _navigationBar.titleTextAttributes = self.titleTextColor ? @{
+                                                               NSForegroundColorAttributeName: self.titleTextColor
+                                                               } : nil;
+  
+  RCTFindNavBarShadowViewInView(_navigationBar).hidden = self.shadowHidden;
+}
+
+- (void)setNavigationItem:(UINavigationItem *)navigationItem
+{
+  _navigationItem = navigationItem;
+  _navigationItem.title = self.title;
+  _navigationItem.backBarButtonItem = self.backButtonItem;
+  _navigationItem.leftBarButtonItem = self.leftButtonItem;
+  _navigationItem.rightBarButtonItem = self.rightButtonItem;
+}
+
+- (void)setTitle:(NSString *)title
+{
+  _title = title;
+  _navigationItem.title = title;
+}
 
 - (void)setBackButtonTitle:(NSString *)backButtonTitle
 {
   _backButtonTitle = backButtonTitle;
-  _backButtonItem = nil;
+  if (_navigationItem.backBarButtonItem) {
+    if (_navigationItem.backBarButtonItem.title) {
+      if (_backButtonTitle && _backButtonTitle.length) {
+        _navigationItem.backBarButtonItem.title = _backButtonTitle;
+        return;
+      }
+      _navigationItem.backBarButtonItem = nil;
+    }
+  }
+  [self backButtonItem];
```

Comment:
There are also a lot of redundant checks in this code, even if the logic is correct. There are no null pointer exceptions in Objective-C, so no need to check if things are nil before accessing their sub-properties.

I think it can be simplified to:

  if (_navigationItem.backBarButtonItem.title) {
    // update title if backBarButtonItem already has a title
    _navigationItem.backBarButtonItem.title = _backButtonTitle;
  } else if (_backButtonTitle) {
    // re-create backBarButtonItem
    _navigationItem.backBarButtonItem = nil;
    [self backButtonItem];
  }